### PR TITLE
export live instance config as yaml

### PR DESCRIPTION
**Pull request by ndeloof**
define a Model for the configuration tree (Mapping, Sequence, Scalars) to abstract away from yaml stuff and limit to simplest json-like yaml constructs. 

sample output from a fresh new jenkins instance :
```yaml
jenkins:
  systemMessage: "Jenkins configured automatically by Jenkins Configuration as Code"
  agentProtocols:
  - "CLI-connect"
  - "CLI2-connect"
  - "JNLP-connect"
  - "JNLP2-connect"
  - "JNLP4-connect"
  - "Ping"
  mode: NORMAL
  numExecutors: 0
  scmCheckoutRetryCount: 0
  projectNamingStrategy: "standard"
  disableRememberMe: false
  markupFormatter: "plainText"
  slaveAgentPort: 0
  myViewsTabBar: "standard"
  viewsTabBar: "standard"
```

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/casc/DataBoundConfiguratorTest.java [hunk 1]
```diff
@@ -24,11 +26,11 @@
 
     @Test
     public void configure_databound() throws Exception {
-        Map<String, Object> config = new HashMap<>();
-        config.put("foo", "foo");
-        config.put("bar", true);
-        config.put("qix", 123);
-        config.put("zot", "DataBoundSetter");
+        Mapping config = new Mapping();
+        config.put("foo", new Scalar("foo"));
```
Comment (casz) [inline src/test/java/org/jenkinsci/plugins/casc/DataBoundConfiguratorTest.java]: Why not modify the put<String, String> to convert string to Scalar?

File: src/main/java/org/jenkinsci/plugins/casc/model/Sequence.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package org.jenkinsci.plugins.casc.model;
+
+import java.util.ArrayList;
+
+/**
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+public final class Sequence extends ArrayList<CNode> implements CNode {
```
Comment (casz) [inline src/main/java/org/jenkinsci/plugins/casc/model/Sequence.java]: Would it not be better with more generic collection?

Comment (ndeloof) [reply src/main/java/org/jenkinsci/plugins/casc/model/Sequence.java]: any suggestion ?

File: src/test/java/org/jenkinsci/plugins/casc/DataBoundConfiguratorTest.java (hunk 1 above)
Comment (ndeloof) [reply src/test/java/org/jenkinsci/plugins/casc/DataBoundConfiguratorTest.java]: as Mapping is a Map<String, CNode> java will fail with a cast exception if invoked with anything but a CNode. Also we only use such an explicit `put()` in few test-cases, we prefer to load a sample yaml file for integration tests.

**Comment by ndeloof** [comment]
@casz seems nobody did the yaml map notation used by initial credentials integration.
I guess most people only know yaml as an alternative to json, even it can do way more. Considering this I think it makes sense to limit this to Sequences / Mappins<String,?> / Scalars which should be enough for all reasonable use-cases.

File: src/test/java/org/jenkinsci/plugins/casc/DataBoundConfiguratorTest.java (hunk 1 above)
Comment (ndeloof) [reply src/test/java/org/jenkinsci/plugins/casc/DataBoundConfiguratorTest.java]: .. but we can offer overridden method

Comment (casz) [reply src/test/java/org/jenkinsci/plugins/casc/DataBoundConfiguratorTest.java]: Awesome looks better 👍

File: src/main/java/org/jenkinsci/plugins/casc/model/Sequence.java (hunk 2 above)
Comment (casz) [reply src/main/java/org/jenkinsci/plugins/casc/model/Sequence.java]: For yaml, I guess `ArrayList` or `List` would be fine.

Comment (ndeloof) [reply src/main/java/org/jenkinsci/plugins/casc/model/Sequence.java]: I don't get your note. The goal here is to restrict the model to only sequence of (other supported) CNode types. I selected ArrayList implementation as snakeyaml parser do create instance with a known size.

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/casc/model/Sequence.java]: Right, then dismiss my comment 👍

**Pull request merged**